File: app/src/main/java/com/teambrella/android/ui/teammate/TeammateActivity.java
package com.teambrella.android.ui.teammate;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.google.gson.JsonObject;
import com.teambrella.android.R;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.model.json.JsonWrapper;
import com.teambrella.android.api.server.TeambrellaUris;
import com.teambrella.android.data.base.TeambrellaDataFragment;
import com.teambrella.android.data.base.TeambrellaDataPagerFragment;
import com.teambrella.android.image.TeambrellaImageLoader;
import com.teambrella.android.ui.TeambrellaUser;
import com.teambrella.android.ui.base.ADataHostActivity;
import com.teambrella.android.ui.base.ADataProgressFragment;
import com.teambrella.android.ui.chat.ChatActivity;
import com.teambrella.android.ui.votes.AllVotesActivity;

import io.reactivex.Notification;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;

/**
 * Teammate screen.
 */
public class TeammateActivity extends ADataHostActivity implements ITeammateActivity {

    private static final String TEAMMATE_URI = "teammate_uri";
    private static final String TEAMMATE_NAME = "teammate_name";
    private static final String TEAMMATE_PICTURE = "teammate_picture";
    private static final String TEAMMATE_USER_ID = "teammate_user_id";
    private static final String CURRENCY = "currency";
    private static final String TEAM_ID = "team_id";

    private static final String DATA_FRAGMENT = "data";
    private static final String VOTE_FRAGMENT = "vote";
    private static final String PROXY_FRAGMENT = "proxy";
    private static final String UI_FRAGMENT = "ui";


    private Disposable mDisposal;
    private int mTeammateId = -1;
    private String mUserId = null;
    private String mUserName = null;
    private int mTeamId;
    private Uri mAvatar = null;
    private String mCurrency;
    private Snackbar mSnackBar;
    private TextView mTitleView;


    public static Intent getIntent(Context context, int teamId, String userId, String name, String userPictureUri) {
        return new Intent(context, TeammateActivity.class)
                .putExtra(TEAMMATE_USER_ID, userId)
                .putExtra(TEAMMATE_URI, TeambrellaUris.getTeammateUri(teamId, userId))
                .putExtra(TEAMMATE_NAME, name)
                .putExtra(TEAM_ID, teamId)
                .putExtra(TEAMMATE_PICTURE, userPictureUri);
    }

    public static void start(Context context, int teamId, String userId, String name, String userPictureUri) {
        context.startActivity(getIntent(context, teamId, userId, name, userPictureUri));
    }


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        mCurrency = getIntent().getStringExtra(CURRENCY);
        mUserId = getIntent().getStringExtra(TEAMMATE_USER_ID);
        mTeamId = getIntent().getIntExtra(TEAM_ID, 0);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activiity_teammate);


        FragmentManager fragmentManager = getSupportFragmentManager();

        if (fragmentManager.findFragmentByTag(UI_FRAGMENT) == null) {
            fragmentManager.beginTransaction()
                    .add(R.id.container, ADataProgressFragment.getInstance(new String[]{DATA_FRAGMENT, VOTE_FRAGMENT, PROXY_FRAGMENT}, TeammateFragment.class), UI_FRAGMENT)
                    .commit();
        }

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeAsUpIndicator(R.drawable.ic_arrow_back_vector);
            actionBar.setDisplayOptions(actionBar.getDisplayOptions() | ActionBar.DISPLAY_SHOW_CUSTOM);
            actionBar.setCustomView(R.layout.teammate_toolbar_view);
            View customView = actionBar.getCustomView();
            Toolbar parent = (Toolbar) customView.getParent();
            parent.setPadding(0, 0, 0, 0);
            parent.setContentInsetsAbsolute(0, 0);
            mTitleView = customView.findViewById(R.id.title);


            View sendMessageView = customView.findViewById(R.id.send_message);
            if (TeambrellaUser.get(this).getUserId().equals(mUserId)) {
                sendMessageView.setVisibility(View.GONE);
            } else {
                customView.findViewById(R.id.send_message).setOnClickListener(v ->
                        ChatActivity.startConversationChat(this, mUserId, mUserName, mAvatar));
            }

        }
        setTitle(getIntent().getStringExtra(TEAMMATE_NAME));
    }

    @Override
    protected void onStart() {
        super.onStart();
        mDisposal = getObservable(DATA_FRAGMENT).subscribe(this::onDataUpdated);
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (mDisposal != null && !mDisposal.isDisposed()) {
            mDisposal.dispose();
        }
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        //getMenuInflater().inflate(R.menu.votes, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            case R.id.votes:
                AllVotesActivity.startTeammateAllVotes(this, mTeamId, mTeammateId);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }


    @Override
    public void setTitle(CharSequence title) {
        mTitleView.setText(title);
    }


    @Override
    public void postVote(double vote) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        TeambrellaDataFragment dataFragment = (TeambrellaDataFragment) fragmentManager.findFragmentByTag(VOTE_FRAGMENT);
        if (dataFragment != null) {
            dataFragment.load(TeambrellaUris.getTeammateVoteUri(mTeammateId, vote));
        }
    }

    @Override
    public void setAsProxy(boolean set) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        TeambrellaDataFragment dataFragment = (TeambrellaDataFragment) fragmentManager.findFragmentByTag(VOTE_FRAGMENT);
        if (dataFragment != null) {
            dataFragment.load(TeambrellaUris.setMyProxyUri(mUserId, set));
        }
    }


    @Override
    public boolean isItMe() {
        return mUserId != null && mUserId.equals(TeambrellaUser.get(this).getUserId());
    }

    @Override
    protected String[] getDataTags() {
        return new String[]{DATA_FRAGMENT, VOTE_FRAGMENT, PROXY_FRAGMENT};
    }

    @Override
    protected String[] getPagerTags() {
        return new String[]{};
    }

    @Override
    protected TeambrellaDataPagerFragment getDataPagerFragment(String tag) {
        return null;
    }


    @Override
    public String getCurrency() {
        return mCurrency;
    }


    @Override
    public int getTeamId() {
        return mTeamId;
    }

    @Override
    public int getTeammateId() {
        return mTeammateId;
    }

    @Override
    public void showSnackBar(@StringRes int text) {
        if (mSnackBar == null) {
            mSnackBar = Snackbar.make(findViewById(R.id.container), text, Snackbar.LENGTH_LONG);

            mSnackBar.addCallback(new Snackbar.Callback() {
                @Override
                public void onShown(Snackbar sb) {
                    super.onShown(sb);
                }

                @Override
                public void onDismissed(Snackbar transientBottomBar, int event) {
                    super.onDismissed(transientBottomBar, event);
                    mSnackBar = null;
                }
            });

            mSnackBar.show();
        }
    }


    @Override
    protected TeambrellaDataFragment getDataFragment(String tag) {
        switch (tag) {
            case DATA_FRAGMENT:
                return TeambrellaDataFragment
                        .getInstance(getIntent().getParcelableExtra(TEAMMATE_URI));
            case VOTE_FRAGMENT:
            case PROXY_FRAGMENT:
                return TeambrellaDataFragment.getInstance(null);
        }
        return null;
    }


    private void onDataUpdated(Notification<JsonObject> notification) {
        if (notification.isOnNext()) {
            Observable.fromArray(notification.getValue())
                    .map(JsonWrapper::new)
                    .map(node -> node.getObject(TeambrellaModel.ATTR_DATA))
                    .doOnNext(node -> mTeammateId = node.getInt(TeambrellaModel.ATTR_DATA_ID))
                    .map(node -> node.getObject(TeambrellaModel.ATTR_DATA_ONE_BASIC))
                    .doOnNext(node -> mUserId = node.getString(TeambrellaModel.ATTR_DATA_USER_ID))
                    .doOnNext(node -> mUserName = node.getString(TeambrellaModel.ATTR_DATA_NAME))
                    .doOnNext(node -> mAvatar = TeambrellaImageLoader.getImageUri(node.getString(TeambrellaModel.ATTR_DATA_AVATAR)))
                    .onErrorReturnItem(new JsonWrapper(null))
                    .blockingFirst();
        }
    }


}

File: app/src/main/java/com/teambrella/android/ui/teammate/TeammateActivity.kt
package com.teambrella.android.ui.teammate

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.ActionBar
import android.support.v7.widget.Toolbar
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import com.google.gson.JsonObject
import com.teambrella.android.R
import com.teambrella.android.api.*
import com.teambrella.android.api.server.TeambrellaUris
import com.teambrella.android.data.base.subscribeAutoDispose
import com.teambrella.android.services.TeambrellaNotificationServiceClient
import com.teambrella.android.services.push.INotificationMessage
import com.teambrella.android.ui.base.*
import com.teambrella.android.ui.chat.ChatActivity
import com.teambrella.android.ui.dialog.ProgressDialogFragment
import com.teambrella.android.util.ImagePicker
import io.reactivex.Notification

fun getTeammateIntent(context: Context, teamId: Int, userId: String, name: String?, userPictureUri: String?) = Intent(context, TeammateActivity::class.java).apply {
    this.userId = userId
    this.teamId = teamId
    this.userName = name
    this.avatar = userPictureUri
    this.uri = TeambrellaUris.getTeammateUri(teamId, userId)
}

fun startTeammateActivity(context: Context, teamId: Int, userId: String, name: String?, userPictureUri: String?) {
    context.startActivity(getTeammateIntent(context, teamId, userId, name, userPictureUri))
}

class TeammateActivity : ATeambrellaActivity(), ITeammateActivity {

    companion object {
        private const val TEAMMATE = "teammate"
        private const val VOTE = "vote"
        private const val PROXY = "proxy"
        private const val UI = "ui"
        private const val PLEASE_WAIT_DIALOG_FRAGMENT_TAG = "please_wait_tag"
    }

    override val dataTags: Array<String> = arrayOf(TEAMMATE, VOTE, PROXY)

    private var userId: String? = null
        get() = field ?: intent.userId

    private var teammateId: Int? = null

    private var userName: String? = null
        get() = field ?: intent.userName

    private var avatar: String? = null
        get() = field ?: intent.avatar

    private var topicId: String? = null

    private var mSnackBar: Snackbar? = null

    private lateinit var teambrellaBroadcastManager: TeambrellaBroadcastManager
    private lateinit var titleView: TextView
    private lateinit var notificationClient: TeammateNotificationClient
    private lateinit var imagePicker: ImagePicker
    private var showPleaseWaitOnResume = false


    override fun onCreate(savedInstanceState: Bundle?) {
        teambrellaBroadcastManager = TeambrellaBroadcastManager(this)
        teambrellaBroadcastManager.registerReceiver(chatBroadCastReceiver)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_teammate)

        supportFragmentManager?.apply {
            if (findFragmentByTag(UI) == null) {
                beginTransaction()
                        .add(R.id.container, createDataFragment(arrayOf(TEAMMATE, VOTE, PROXY), TeammateFragment::class.java), UI)
                        .commit()
            }
        }

        supportActionBar?.apply {

            setDisplayHomeAsUpEnabled(true)
            setHomeAsUpIndicator(R.drawable.ic_arrow_back_vector)
            displayOptions = displayOptions or ActionBar.DISPLAY_SHOW_CUSTOM

            setCustomView(R.layout.teammate_toolbar_view)

            customView.apply {
                (parent as Toolbar).apply {
                    setPadding(0, 0, 0, 0)
                    setContentInsetsAbsolute(0, 0)
                }
            }

            titleView = customView.findViewById<TextView>(R.id.title).apply {
                text = intent?.userName
            }

            customView.findViewById<View>(R.id.send_message)?.apply {
                visibility = View.GONE
            }
        }

        notificationClient = TeammateNotificationClient(this)
        notificationClient.connect()

        getObservable(TEAMMATE).observe(this, Observer { notification ->
            this@TeammateActivity.onDataUpdated(notification ?: throw kotlin.RuntimeException())
        })


        imagePicker = ImagePicker(this)

    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : TeambrellaDataViewModel> getDataViewModelClass(tag: String): Class<T>? = when (tag) {
        TEAMMATE -> TeammateViewModel::class.java as Class<T>
        else -> super.getDataViewModelClass(tag)
    }

    override fun getDataConfig(tag: String): Bundle? = when (tag) {
        TEAMMATE -> Bundle().apply { this.uri = intent?.uri }
        VOTE, PROXY -> Bundle()
        else -> super.getDataConfig(tag)
    }

    override fun setTitle(title: CharSequence?) {
        titleView.text = title
    }


    override fun onResume() {
        super.onResume()
        notificationClient.onPause()
        if (showPleaseWaitOnResume) {
            showPleaseWaitOnResume = false
            showPleaseWait()
        }
    }

    override fun onPause() {
        super.onPause()
        notificationClient.onResume()
    }


    override fun onDestroy() {
        super.onDestroy()
        notificationClient.disconnect()
        teambrellaBroadcastManager.unregisterReceiver(chatBroadCastReceiver)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        imagePicker.onActivityResult(requestCode, resultCode, data)?.apply {
            showPleaseWaitOnResume = true
            subscribeAutoDispose({ t ->
                request(TeambrellaUris.setAvatarUri(t.file.absolutePath))
            }, {
                showPleaseWaitOnResume = false
                hidePleaseWait()
            }, { })
        }
    }

    private fun onDataUpdated(notification: Notification<JsonObject>) {
        if (notification.isOnNext) {
            notification.value.data.let { _data ->
                teammateId = _data.intId
                _data.basic?.let { _basic ->
                    userId = _basic.userId
                    userName = _basic.userName
                    avatar = _basic.avatar
                }


                _data.team?.let { _team ->
                    supportActionBar?.customView?.findViewById<View>(R.id.send_message)?.apply {
                        when {
                            user.userId == intent.userId ||
                                    _team.teamAccessLevel == TeambrellaModel.TeamAccessLevel.READ_ONLY_ALL_AND_STEALTH
                            -> {
                                visibility = View.GONE
                            }
                            else -> {
                                visibility = View.VISIBLE
                                setOnClickListener {
                                    ChatActivity.startConversationChat(this@TeammateActivity, userId, userName, avatar)
                                }
                            }
                        }
                    }
                }

                _data.discussionPart?.let { _discussion ->
                    topicId = _discussion.topicId
                }
            }

        }
    }

    override fun postVote(vote: Double) {
        teammateId?.let {
            ViewModelProviders.of(this).get(VOTE, TeambrellaDataViewModel::class.java)
                    .load(TeambrellaUris.getTeammateVoteUri(it, vote))
        }
    }

    override fun setAsProxy(set: Boolean) {
        userId?.let {
            ViewModelProviders.of(this).get(PROXY, TeambrellaDataViewModel::class.java)
                    .load(TeambrellaUris.getSetMyProxyUri(it, set))
        }
    }

    override fun isItMe(): Boolean = (userId == user.userId)


    override fun getCurrency(): String? = null

    override fun showSnackBar(text: Int) {
        if (mSnackBar == null) {
            mSnackBar = Snackbar.make(findViewById(R.id.container), text, Snackbar.LENGTH_LONG)
            mSnackBar?.addCallback(object : Snackbar.Callback() {
                override fun onDismissed(transientBottomBar: Snackbar?, event: Int) {
                    super.onDismissed(transientBottomBar, event)
                    mSnackBar = null
                }
            })
            mSnackBar?.show()
        }
    }

    override fun getTeamId(): Int = intent.teamId

    override fun getTeammateName(): String? = userName

    override fun getTeammateId(): Int = teammateId ?: 0

    override val isRequestable = true

    override fun launchActivity(intent: Intent?) {
        startActivityForResult(intent, 3)
    }

    override fun setAvatar() {
        imagePicker.startPicking(getString(R.string.select_profile_photo))
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        imagePicker.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onRequestResult(response: Notification<JsonObject>) {
        super.onRequestResult(response)
        response.takeIf { it.isOnNext }?.value?.status?.uri?.let {
            when (TeambrellaUris.sUriMatcher.match(Uri.parse(it))) {
                TeambrellaUris.SET_AVATAR -> {
                    showPleaseWaitOnResume = false
                    hidePleaseWait()
                    load(TEAMMATE)
                }
            }
        }
    }

    private inner class TeammateNotificationClient(context: Context) : TeambrellaNotificationServiceClient(context) {

        private var mResumed: Boolean = false
        private var mUpdateOnResume: Boolean = false

        override fun onPushMessage(message: INotificationMessage): Boolean {
            val topicId = message.topicId
            if (topicId != null && topicId == this@TeammateActivity.topicId) {
                if (mResumed) {
                    load(TEAMMATE)
                } else {
                    mUpdateOnResume = true
                }
            }
            return false
        }

        internal fun onPause() {
            mResumed = false
        }

        internal fun onResume() {
            mResumed = true
            if (mUpdateOnResume) {
                load(TEAMMATE)
                mUpdateOnResume = false
            }
        }
    }


    private fun showPleaseWait() {
        val fragmentManager = supportFragmentManager
        if (fragmentManager.findFragmentByTag(PLEASE_WAIT_DIALOG_FRAGMENT_TAG) == null) {
            ProgressDialogFragment().show(fragmentManager, PLEASE_WAIT_DIALOG_FRAGMENT_TAG)
        }
    }

    private fun hidePleaseWait() {
        val fragmentManager = supportFragmentManager
        val fragment = fragmentManager.findFragmentByTag(PLEASE_WAIT_DIALOG_FRAGMENT_TAG)
        if (fragment != null && fragment is ProgressDialogFragment) {
            fragment.dismiss()
        }
    }


    private val chatBroadCastReceiver = object : TeambrellaBroadcastReceiver() {
        override fun onTopicRead(topicId: String) {
            super.onTopicRead(topicId)
            ViewModelProviders.of(this@TeammateActivity)
                    .get(TEAMMATE, TeammateViewModel::class.java)
                    .markTopicRead(topicId)
        }
    }
}